src/section/section_chain.rs

```diff
@@ -0,0 +1,1114 @@
+                (&block.key, Some(block.parent_index))
+            } else {
+                (&self.root, None)
+            };
+
+            if trusted_keys.contains(key) {
+                return true;
+            }
+
+            if let Some(next_index) = parent_index {
+                index = next_index;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    /// Compare the two keys by their position in the chain. The key that is higher (closer to the
+    /// last key) is considered `Greater`. If exactly one of the keys is not in the chain, the other
+    /// one is implicitly considered `Greater`. If none are in the chain, they are considered
+    /// `Equal`.
+    pub fn cmp_by_position(&self, lhs: &bls::PublicKey, rhs: &bls::PublicKey) -> Ordering {
+        match (self.index_of(lhs), self.index_of(rhs)) {
+            (Some(lhs), Some(rhs)) => lhs.cmp(&rhs),
+            (Some(_), None) => Ordering::Greater,
+            (None, Some(_)) => Ordering::Less,
+            (None, None) => Ordering::Equal,
+        }
+    }
+
+    /// Returns the number of blocks in the chain. This is always >= 1.
+    pub fn len(&self) -> usize {
+        1 + self.tree.len()
+    }
+
+    fn insert_block(&mut self, new_block: Block) -> usize {
+        // Find the index into `self.tree` to insert the new key at. All the keys above will be
+        // pushed upwards.
+        let insert_at = self
+            .tree
+            .iter()
+            .enumerate()
+            .skip(new_block.parent_index)
+            .find(|(_, block)| {
+                block.parent_index != new_block.parent_index || block.key >= new_block.key
+            })
+            .map(|(index, _)| index)
+            .unwrap_or(self.tree.len());
+
+        // If the key already exists in the chain, do nothing but still return success to make the
+        // `insert` operation idempotent.
+        if self.tree.get(insert_at).map(|block| &block.key) != Some(&new_block.key) {
+            self.tree.insert(insert_at, new_block);
+
+            // Adjust the parent indices of the keys whose parents are above the inserted key.
+            for block in &mut self.tree {
+                if block.parent_index > insert_at {
+                    block.parent_index += 1;
+                }
+            }
+        }
+
+        insert_at + 1
+    }
+
+    fn index_of(&self, key: &bls::PublicKey) -> Option<usize> {
+        self.keys()
+            .rev()
+            .position(|existing_key| existing_key == key)
+            .map(|rev_position| self.len() - rev_position - 1)
+    }
+
+    fn parent_index_at(&self, index: usize) -> Option<usize> {
+        if index == 0 {
+            None
+        } else {
+            self.tree.get(index - 1).map(|block| block.parent_index)
+        }
+    }
+
+    // Is the key at `lhs` an ancestor of the key at `rhs`?
+    fn is_ancestor(&self, lhs: usize, rhs: usize) -> bool {
+        let mut index = rhs;
```

Comment:
what if lhs is 0, i.e. does it mean the root ?

Reply:
Yes, that should work as expected.